File: study/examples/Algorithm/python/01_complexity.py
"""
시간 복잡도 비교 실험
Time Complexity Comparison

다양한 시간 복잡도를 가진 알고리즘들의 실행 시간을 비교합니다.
"""

import time
import random


def measure_time(func, *args):
    """함수 실행 시간 측정"""
    start = time.perf_counter()
    result = func(*args)
    end = time.perf_counter()
    return end - start, result


# =============================================================================
# O(1) - 상수 시간
# =============================================================================
def constant_time(arr):
    """배열의 첫 번째 요소 반환 - O(1)"""
    if arr:
        return arr[0]
    return None


# =============================================================================
# O(log n) - 로그 시간
# =============================================================================
def binary_search(arr, target):
    """이분 탐색 - O(log n)"""
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1


# =============================================================================
# O(n) - 선형 시간
# =============================================================================
def linear_search(arr, target):
    """선형 탐색 - O(n)"""
    for i, val in enumerate(arr):
        if val == target:
            return i
    return -1


def find_max(arr):
    """최댓값 찾기 - O(n)"""
    if not arr:
        return None
    max_val = arr[0]
    for val in arr:
        if val > max_val:
            max_val = val
    return max_val


# =============================================================================
# O(n log n) - 선형 로그 시간
# =============================================================================
def merge_sort(arr):
    """병합 정렬 - O(n log n)"""
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])

    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


# =============================================================================
# O(n²) - 제곱 시간
# =============================================================================
def bubble_sort(arr):
    """버블 정렬 - O(n²)"""
    arr = arr.copy()
    n = len(arr)
    for i in range(n):
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr


def has_duplicate_naive(arr):
    """중복 검사 (naive) - O(n²)"""
    n = len(arr)
    for i in range(n):
        for j in range(i + 1, n):
            if arr[i] == arr[j]:
                return True
    return False


# =============================================================================
# O(2^n) - 지수 시간
# =============================================================================
def fibonacci_recursive(n):
    """피보나치 (재귀) - O(2^n)"""
    if n <= 1:
        return n
    return fibonacci_recursive(n - 1) + fibonacci_recursive(n - 2)


def fibonacci_dp(n):
    """피보나치 (DP) - O(n)"""
    if n <= 1:
        return n
    dp = [0] * (n + 1)
    dp[1] = 1
    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
    return dp[n]


# =============================================================================
# 실험 및 결과 출력
# =============================================================================
def run_experiments():
    """시간 복잡도 실험 실행"""
    print("=" * 60)
    print("시간 복잡도 비교 실험")
    print("=" * 60)

    # 데이터 준비
    sizes = [100, 1000, 10000]

    for size in sizes:
        print(f"\n[ 배열 크기: {size} ]")
        print("-" * 40)

        arr = list(range(size))
        random_arr = random.sample(range(size * 10), size)
        target = arr[size // 2]  # 중간 값

        # O(1) 테스트
        t, _ = measure_time(constant_time, arr)
        print(f"O(1)     상수 시간:      {t * 1000:.6f} ms")

        # O(log n) 테스트
        t, _ = measure_time(binary_search, arr, target)
        print(f"O(log n) 이분 탐색:      {t * 1000:.6f} ms")

        # O(n) 테스트
        t, _ = measure_time(linear_search, arr, target)
        print(f"O(n)     선형 탐색:      {t * 1000:.6f} ms")

        t, _ = measure_time(find_max, random_arr)
        print(f"O(n)     최댓값 찾기:    {t * 1000:.6f} ms")

        # O(n log n) 테스트
        if size <= 10000:
            t, _ = measure_time(merge_sort, random_arr)
            print(f"O(n log n) 병합 정렬:  {t * 1000:.6f} ms")

        # O(n²) 테스트 (작은 크기만)
        if size <= 1000:
            t, _ = measure_time(bubble_sort, random_arr)
            print(f"O(n²)    버블 정렬:      {t * 1000:.6f} ms")

    # O(2^n) vs O(n) 피보나치 비교
    print("\n[ 피보나치 비교: O(2^n) vs O(n) ]")
    print("-" * 40)

    for n in [10, 20, 30]:
        t_recursive, _ = measure_time(fibonacci_recursive, n)
        t_dp, _ = measure_time(fibonacci_dp, n)
        print(f"n={n}: 재귀 O(2^n) = {t_recursive * 1000:.4f} ms, "
              f"DP O(n) = {t_dp * 1000:.6f} ms")

    print("\n" + "=" * 60)
    print("실험 완료!")
    print("=" * 60)


if __name__ == "__main__":
    run_experiments()

File: study/examples/Algorithm/cpp/01_complexity.cpp
/*
 * 시간 복잡도 (Time Complexity)
 * Big O Notation and Complexity Analysis
 *
 * 알고리즘의 효율성을 분석하는 방법입니다.
 */

#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <functional>

using namespace std;

// =============================================================================
// 1. O(1) - 상수 시간
// =============================================================================

int constantTime(const vector<int>& arr) {
    // 배열 크기와 무관하게 항상 일정한 시간
    return arr[0];
}

// =============================================================================
// 2. O(log n) - 로그 시간
// =============================================================================

int binarySearch(const vector<int>& arr, int target) {
    int left = 0, right = arr.size() - 1;

    while (left <= right) {
        int mid = left + (right - left) / 2;

        if (arr[mid] == target)
            return mid;
        else if (arr[mid] < target)
            left = mid + 1;
        else
            right = mid - 1;
    }

    return -1;
}

// =============================================================================
// 3. O(n) - 선형 시간
// =============================================================================

int linearSearch(const vector<int>& arr, int target) {
    for (size_t i = 0; i < arr.size(); i++) {
        if (arr[i] == target)
            return i;
    }
    return -1;
}

int sumArray(const vector<int>& arr) {
    int sum = 0;
    for (int x : arr) {
        sum += x;
    }
    return sum;
}

// =============================================================================
// 4. O(n log n) - 선형 로그 시간
// =============================================================================

void merge(vector<int>& arr, int left, int mid, int right) {
    vector<int> L(arr.begin() + left, arr.begin() + mid + 1);
    vector<int> R(arr.begin() + mid + 1, arr.begin() + right + 1);

    size_t i = 0, j = 0;
    int k = left;

    while (i < L.size() && j < R.size()) {
        if (L[i] <= R[j])
            arr[k++] = L[i++];
        else
            arr[k++] = R[j++];
    }

    while (i < L.size()) arr[k++] = L[i++];
    while (j < R.size()) arr[k++] = R[j++];
}

void mergeSort(vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

// =============================================================================
// 5. O(n²) - 이차 시간
// =============================================================================

void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

int countPairs(const vector<int>& arr) {
    int count = 0;
    int n = arr.size();
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            count++;
        }
    }
    return count;  // n*(n-1)/2
}

// =============================================================================
// 6. O(2^n) - 지수 시간
// =============================================================================

int fibonacciRecursive(int n) {
    if (n <= 1) return n;
    return fibonacciRecursive(n - 1) + fibonacciRecursive(n - 2);
}

// O(n) 최적화 버전
int fibonacciIterative(int n) {
    if (n <= 1) return n;

    int prev2 = 0, prev1 = 1;
    for (int i = 2; i <= n; i++) {
        int curr = prev1 + prev2;
        prev2 = prev1;
        prev1 = curr;
    }
    return prev1;
}

// =============================================================================
// 7. 시간 측정 유틸리티
// =============================================================================

template<typename Func, typename... Args>
double measureTime(Func func, Args&&... args) {
    auto start = chrono::high_resolution_clock::now();
    func(forward<Args>(args)...);
    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> diff = end - start;
    return diff.count();
}

// =============================================================================
// 8. 공간 복잡도 예제
// =============================================================================

// O(1) 공간
void reverseInPlace(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n / 2; i++) {
        swap(arr[i], arr[n - 1 - i]);
    }
}

// O(n) 공간
vector<int> reverseWithCopy(const vector<int>& arr) {
    vector<int> result(arr.rbegin(), arr.rend());
    return result;
}

// =============================================================================
// 테스트
// =============================================================================

void printArray(const vector<int>& arr, int limit = 10) {
    cout << "    [";
    for (size_t i = 0; i < arr.size() && i < (size_t)limit; i++) {
        cout << arr[i];
        if (i < arr.size() - 1 && i < (size_t)limit - 1) cout << ", ";
    }
    if (arr.size() > (size_t)limit) cout << ", ...";
    cout << "]" << endl;
}

int main() {
    cout << "============================================================" << endl;
    cout << "시간 복잡도 (Time Complexity) 예제" << endl;
    cout << "============================================================" << endl;

    // 1. O(1)
    cout << "\n[1] O(1) - 상수 시간" << endl;
    vector<int> arr1 = {5, 2, 8, 1, 9};
    cout << "    첫 번째 원소: " << constantTime(arr1) << endl;

    // 2. O(log n)
    cout << "\n[2] O(log n) - 이분 탐색" << endl;
    vector<int> arr2 = {1, 3, 5, 7, 9, 11, 13, 15};
    int idx = binarySearch(arr2, 7);
    cout << "    배열: [1,3,5,7,9,11,13,15]" << endl;
    cout << "    7의 위치: " << idx << endl;

    // 3. O(n)
    cout << "\n[3] O(n) - 선형 탐색" << endl;
    vector<int> arr3 = {4, 2, 7, 1, 9, 3};
    cout << "    배열 합: " << sumArray(arr3) << endl;

    // 4. O(n log n)
    cout << "\n[4] O(n log n) - 병합 정렬" << endl;
    vector<int> arr4 = {64, 34, 25, 12, 22, 11, 90};
    cout << "    정렬 전: ";
    printArray(arr4);
    mergeSort(arr4, 0, arr4.size() - 1);
    cout << "    정렬 후: ";
    printArray(arr4);

    // 5. O(n²)
    cout << "\n[5] O(n²) - 버블 정렬" << endl;
    vector<int> arr5 = {64, 34, 25, 12, 22, 11, 90};
    bubbleSort(arr5);
    cout << "    정렬 후: ";
    printArray(arr5);
    cout << "    5개 원소의 쌍 개수: " << countPairs(vector<int>(5)) << endl;

    // 6. O(2^n) vs O(n)
    cout << "\n[6] O(2^n) vs O(n) - 피보나치" << endl;
    cout << "    피보나치(20) 재귀: " << fibonacciRecursive(20) << endl;
    cout << "    피보나치(20) 반복: " << fibonacciIterative(20) << endl;
    cout << "    피보나치(40) 반복: " << fibonacciIterative(40) << endl;

    // 7. 공간 복잡도
    cout << "\n[7] 공간 복잡도" << endl;
    vector<int> arr7 = {1, 2, 3, 4, 5};
    cout << "    원본: ";
    printArray(arr7);
    reverseInPlace(arr7);
    cout << "    O(1) 공간 뒤집기: ";
    printArray(arr7);

    // 8. 복잡도 요약
    cout << "\n[8] 복잡도 요약" << endl;
    cout << "    | 복잡도    | 1000개 연산 | 예시              |" << endl;
    cout << "    |-----------|-------------|-------------------|" << endl;
    cout << "    | O(1)      | 1           | 배열 인덱싱       |" << endl;
    cout << "    | O(log n)  | 10          | 이분 탐색         |" << endl;
    cout << "    | O(n)      | 1000        | 선형 탐색         |" << endl;
    cout << "    | O(n log n)| 10000       | 병합 정렬         |" << endl;
    cout << "    | O(n²)     | 1000000     | 버블 정렬         |" << endl;
    cout << "    | O(2^n)    | 매우 큼     | 모든 부분집합     |" << endl;

    cout << "\n============================================================" << endl;

    return 0;
}
